iwyu.cc

```diff
@@ -2006,28 +2006,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // compiler a fighting chance of generating correct code.
   bool VisitCastExpr(clang::CastExpr* expr) {
     if (CanIgnoreCurrentASTNode())  return true;
-    const Type* const from_type = GetTypeOf(expr->getSubExprAsWritten());
-    const Type* const to_type = GetTypeOf(expr);
-    const Type* const deref_from_type = RemovePointersAndReferences(from_type);
-    const Type* const deref_to_type = RemovePointersAndReferences(to_type);
-
-    // For all those casts that don't result in function calls
-    // (everything except a user-defined cast or a constructor cast),
-    // we only care about the need for full types when casting either
-    // a pointer to a pointer, or any type to a reference.
-    // Unfortunately, when casting to a reference, clang seems to
-    // strip the reference off of to_type, so we need a separate
-    // function call to tell.
-    if (expr->getCastKind() != clang::CK_UserDefinedConversion &&
-        expr->getCastKind() != clang::CK_ConstructorConversion) {
-      if (!((from_type->hasPointerRepresentation() &&   // pointer or reference
-             to_type->hasPointerRepresentation()) ||
-            IsCastToReferenceType(expr)))
-        return true;    // we only care about ptr-to-ptr casts for this check
+    const Type* from_type = GetTypeOf(expr->getSubExprAsWritten());
+    const Type* to_type = GetTypeOf(expr);
+
+    // If this cast requires a conversion function, look up its return type so
+    // we can see through up/down casts via user-defined conversions.
+    const Type* converted_to_type = nullptr;
+    if (const NamedDecl* conv_decl = expr->getConversionFunction()) {
+      converted_to_type =
+          cast<FunctionDecl>(conv_decl)->getReturnType().getTypePtr();
     }
```

Comment:
The *if* block can be moved inside the `clang::.+DerivedToBase.+` cases below. That's the only place where it is needed.

Reply:
I suspect that more cases will need the converted-to-type (e.g. the dynamic-casts), but I haven't bothered trying to come up with testcases for it.

It makes intuitive sense to me to keep the from-type, to-type and converted-to-type resolution in the same place, because they represent the triad of types we need to reason about a cast expression (depending on which kind of cast it is).

So I'd rather keep it out here, but can move it if you insist.